src/QldbSession.ts

```diff
@@ -12,136 +12,82 @@
-                    warn(`OCC conflict or retriable exception occurred: ${e}.`);
-                } else {
-                    throw e;
-                }
-            }
-            executionContext.incrementExecutionAttempt();
-            await this._retrySleep(executionContext, retryConfig, transaction);
+    async endSession(): Promise<void> {
+        try {
+            this._isAlive = false;
+            await this._communicator.endSession();
+        } catch (e) {
+            // We will only log issues ending the session, as QLDB will clean them after a timeout.
+            warn(`Errors ending session: ${e}.`);
         }
     }
 
-
-    getSessionToken(): string {
-        return this._communicator.getSessionToken();
-    }
-
-    isSessionOpen(): Boolean {
-        return !this._isClosed;
-    }
-
-    async startTransaction(): Promise<Transaction> {
+    async executeLambda<Type>(
+        transactionLambda: (transactionExecutor: TransactionExecutor) => Promise<Type>
+    ): Promise<Type> {
+        let transaction: Transaction;
+        let transactionId: string = null;
         try {
-            const startTransactionResult: StartTransactionResult = await this._communicator.startTransaction();
-            const transaction: Transaction = new Transaction(
-                this._communicator,
-                startTransactionResult.TransactionId
-            );
-            return transaction;
+            transaction = await this._startTransaction();
+            transactionId = transaction.getTransactionId();
+            const executor: TransactionExecutor = new TransactionExecutor(transaction);
+            const returnedValue: Type = await transactionLambda(executor);
+            await transaction.commit();
+            return returnedValue;
         } catch (e) {
-            if (isBadRequestException(e)) {
-                throw new StartTransactionError(e);
+            const isRetriable: boolean = isRetriableException(e);
+            const isISE: boolean = isInvalidSessionException(e);
+            if (isISE && !isTransactionExpiredException(e)) {
+                // Underlying session is dead on InvalidSessionException except for transaction expiry
+                this._isAlive = false;
+            } else if (!isOccConflictException(e)) {
+                // OCC does not need session state reset as the transaction is implicitly closed
+                await this._tryAbort();
             }
-            throw e;
+            throw new ExecuteError(e, isRetriable, isISE, transactionId);
         }
     }
 
-    private async _noThrowAbort(transaction: Transaction): Promise<void> {
+    async _startTransaction(): Promise<Transaction> {
+        const startTransactionResult: StartTransactionResult = await this._communicator.startTransaction();
+        return new Transaction(this._communicator, startTransactionResult.TransactionId);
+    }
+
+    private async _tryAbort(): Promise<void> {
```

Comment:
Generally a tryX function would return true or false depending on if it was successful at the try or not.